network/src/connection_manager.rs

```diff
@@ -0,0 +1,404 @@
+        // self.channels.get(remote_address)
+
+        // Acquire the tmp_connections read lock.
+        let tmp_connections = self.tmp_connections.read().await;
+        // Forget the local address.
+        let channel = tmp_connections.get(remote_address);
+        // Drop the tmp_connections read lock.
+        drop(tmp_connections);
+        channel
+    }
+
+    /// Stores a new channel at the peer address it is connected to.
+    pub fn add_channel(&mut self, channel: &Arc<Channel>) {
+        self.channels.insert(channel.address, channel.clone());
+    }
+
+    /// Manages all peer connections and processes updates with all connected peers.
+    #[inline]
+    pub async fn handler(&self) {
+        // If the node is connected to less peers than the minimum required,
+        // ask every peer the node is connected to for more peers.
+        if self.get_num_connected().await < self.minimum_peer_count {
+            // Broadcast a `GetPeers` message to request for more peers.
+            self.broadcast_getpeers_requests().await;
+            // Attempt a connection request with every disconnected peer.
+            self.connect_to_all_disconnected_peers().await;
+            // Attempt a connection request with every bootnode peer again.
+            // The goal is to reconnect with any bootnode peers we might
+            // have failed to connect to. The manager will filter attempts
+            // to connect with itself or already connected bootnode peers.
+            self.connect_to_bootnodes().await;
+        }
+
+        // TODO (howardwu): Unify `Ping` and `Version`requests.
+        //  This is a remnant and these currently do not need to be distinct.
+
+        // Broadcast a `Ping` request to each connected peer.
+        self.broadcast_ping_requests().await;
+        // Broadcast a `Version` request to each connected peer.
+        self.broadcast_version_requests().await;
+
+        // Store the internal state of the peer book.
+        self.store_peer_book().await;
+    }
+
+    /// An internal operation to write the internal state of the peer book to storage.
+    #[inline]
+    async fn store_peer_book(&self) {
+        // Acquire the peer book write lock.
+        let mut peer_book = self.peer_book.write().await;
+        // Store the peer book into the database.
+        peer_book
+            .store(&self.storage)
+            .unwrap_or_else(|error| debug!("Failed to store connected peers in database {}", error));
+        // Drop the peer book write lock.
+        drop(peer_book);
+    }
+
+    /// Broadcasts a `Version` message to all connected peers.
+    #[inline]
+    async fn broadcast_version_requests(&self) {
+        // Get the local address of the node.
+        let local_address = self.get_local_address().await;
+
+        // Broadcast a `Version` message to each connected peer for a periodic sync.
+        if self.get_num_connected().await > 0 {
```

Comment:
this check feels unnecessary, since the loop below won't do anything if `get_all_connected` is empty (implying the result of `get_num_connected` being `0`)